docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,103 @@
+# Architecture
+
+An overview of the architecture for various components within the Toolkit.
+
+## Webviews
+
+The current implementation uses Vue 3 with Single File Components (SFCs) for modularity. Each webview
+is bundled into a single file and packaged into the toolkit at release time. Each component is able to
+act independently, however, they must respect the following principles:
+
+1. State can only be stored in a child component if it is not being used for two-way communication (via events)
+2. If there is two-way communication, store state in the parent
+3. Data should flow down, actions should flow up
+
+Be very mindful about state managment; violating these principles will lead to buggy and hard-to-debug software.
+
+### Bundling
+
+Each webview is bundled into a single file to speed up load times as well as isolate the 'web' modules from the 'node' modules. Webview bundles are automatically generated on compilation by targeting `entry.ts` files when located under a `vue` directory. All bundles are placed directly under `dist`.
+
+Generated bundle names map based off their path relative to `src`: `src/foo/vue/bar/entry.ts` -> `fooBarVue.js`
+
+Running the extension in development mode (e.g. via the `Extension` launch task) starts a local server to automatically rebuild and serve webviews in real-time via hot-module reloading.
+
+### Client/Server
+
+The VS Code API restricts our Webviews to a single `postMessage` function. To simplify developing Webviews, we use a basic client/server architecture to handle message passing between the view and the extension.
+
+Webview (frontend) clients can be created via `WebviewClientFactory`. This generates a simple Proxy to send messages to the extension, mapping the function name to the command name. Unique IDs are also generated to stop requests from receiving extraneous responses. It is **highly** recommened to use the [Volar](https://marketplace.visualstudio.com/items?itemName=johnsoncodehk.volar) extension for syntax highlighting and type-checking when working with SFCs.
```

Comment:
* Big "eh" on the tool recommendation. It's a nice-to-have but there's also no guarantee that this is always going to be around, and it doesn't apply to users using other editors.
* As mentioned above, I'm not 100% sold on the 1:1 request:response model. The frontend should be able to receive arbitrary input as long as it comes in a logical format. e.g. log tailing/realtime graphs. The frontend shouldn't have to make a request for each new piece of input.

Reply:
>  The frontend should be able to receive arbitrary input as long as it comes in a logical format. e.g. log tailing/realtime graphs.

This is already supported. Yes at its core it's a request/response model, but that doesn't mean events can't exist.

Reply:
> Big "eh" on the tool recommendation. It's a nice-to-have but there's also no guarantee that this is always going to be around, and it doesn't apply to users using other editors.

In that case it's to their own detriment. I'm adding the recommendation since it's what I know works and will make things _much_ easier. It's also recommended by the Vue framework. They can always find a Vue language server if not using VS Code. If the specific extension disappears then the doc should be updated with whatever the maintainer thinks works best. Not being opinionated leads to confusion.